libnetwork/drivers/bridge/internal/firewaller/stub.go

```diff
@@ -0,0 +1,157 @@
+// FIXME(thaJeztah): remove once we are a module; the go:build directive prevents go from downgrading language version to go1.16:
+//go:build go1.22 && linux
+
+package firewaller
+
+import (
+	"context"
+	"fmt"
+	"net/netip"
+	"slices"
+
+	"github.com/docker/docker/libnetwork/types"
+)
+
+// StubFirewaller implements a Firewaller for unit tests. It just tracks what it's been asked for.
+type StubFirewaller struct {
+	Config
+	Networks map[string]*StubFirewallerNetwork
+	FFD      bool // filter forward drop
+}
+
+func NewStubFirewaller(config Config) *StubFirewaller {
+	return &StubFirewaller{
+		Config:   config,
+		Networks: make(map[string]*StubFirewallerNetwork),
+	}
+}
+
+func (fw *StubFirewaller) NewNetwork(_ context.Context, nc NetworkConfig) (Network, error) {
+	if nw, ok := fw.Networks[nc.IfName]; ok && !nw.Deleted {
+		return nil, fmt.Errorf("StubFirewaller: network with IfName %q already exists", nc.IfName)
+	}
+	nw := &StubFirewallerNetwork{
+		NetworkConfig: nc,
+		Endpoints:     map[stubEndpoint]struct{}{},
+	}
+	fw.Networks[nc.IfName] = nw
+	return nw, nil
+}
+
+func (fw *StubFirewaller) FilterForwardDrop(_ context.Context, _ IPVersion) error {
+	fw.FFD = true
+	return nil
+}
+
+type stubFirewallerLink struct {
+	parentIP netip.Addr
+	childIP  netip.Addr
+	ports    []types.TransportPort
+}
+
+type stubEndpoint struct {
+	addr4 netip.Addr
+	addr6 netip.Addr
+}
+
+type StubFirewallerNetwork struct {
+	NetworkConfig
+	Deleted   bool
+	Endpoints map[stubEndpoint]struct{}
+	Ports     []types.PortBinding
+	Links     []stubFirewallerLink
+}
+
+func (nw *StubFirewallerNetwork) ReapplyNetworkLevelRules(_ context.Context) error {
+	return nil
+}
+
+func (nw *StubFirewallerNetwork) DelNetworkLevelRules(_ context.Context) error {
+	nw.Deleted = true
+	return nil
+}
```

Comment:
Is there a meaningful difference between the rules being applied and deleted versus the rules never being applied in the first place? From a quick glance at the iptabler implementation it looks like `DelNetworkLevelRules` is the inverse of `NewNetwork`. Would it not be simpler and less error-prone to write assertions if `DelNetworkLevelRules` deleted the corresponding entry from the `StubFirewaller.Networks` map?